librad/src/profile.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright © 2019-2021 The Radicle Foundation <hello@radicle.foundation>
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::{env, fs, io, path::Path};
+use uuid::Uuid;
+
+use crate::paths::{project_dirs, Paths};
+
+/// A [`Profile`] provides [`Paths`] scoped by an identifier.
+///
+/// Profiles with different identifiers have distinct paths.
+#[derive(Debug, Clone)]
+pub struct Profile {
+    id: String,
+    paths: Paths,
+}
+
+impl Profile {
+    /// Creates a profile by loading the profile identifier and paths from
+    /// the environment variables or well-known file.
+    ///
+    /// If `RAD_PROFILE` is set, the value is used as the profile identifier
+    /// and system paths, scoped to the identifier, are used. See
+    /// [`Paths::new`] for details.
+    ///
+    /// Otherwise and if the `RAD_HOME` environment variable is set, the value
+    /// is used to construct the profile with [`Profile::from_root`].
+    ///
+    /// Otherwise, the profile identifier is read from a well-known file and
+    /// system paths scoped to the identifier are used. If the file containing
+    /// the identifier  does not exist, it is created and new identifier is
+    /// generated and written to the file.
+    ///
+    /// On Linux, the path to the active profile is
+    /// `$XDG_CONFIG_HOME/radicle/active_profile`.
+    pub fn load() -> Result<Self, io::Error> {
```

Comment:
Might be good to ensure that the profile id is only a path _component_, or otherwise ensure that it doesn't look funny (eg. `../../WORKGROUP`).

Maybe we can also hand out the id as a `Path`, to reduce suprises when it contains newlines or other funky characters.

Reply:
> Might be good to ensure that the profile id is only a path _component_, or otherwise ensure that it doesn't look funny (eg. `../../WORKGROUP`).

Good point. Will add the validation logic.

> Maybe we can also hand out the id as a `Path`, to reduce suprises when it contains newlines or other funky characters.

Are your referring to `Profile::id`? Having it return `Path` might be confusing, especially since it can only be a path segment since the base directories are all different.

Reply:
Ya I was thinking of the "smart" `Display` impl for `Path`. If we sanitise the input properly, I'm ok with `&str`.

Reply:
Addressed in https://github.com/radicle-dev/radicle-link/pull/496/commits/79a81a3177810bf278b8ba73e644bd822aed573f